pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,216 @@
+
+	allNodeNames := make([]string, 0, len(nodeNameToInfo))
+	for name := range nodeNameToInfo {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeNameToInfo[allNodeNames[i]]
+		if node := nodeInfo.Node(); node != nil {
+			// (1) `node` should satisfy incoming pod's NodeSelector/NodeAffinity
+			// (2) All topologyKeys need to be present in `node`
+			if !predicates.PodMatchesNodeSelectorAndAffinityTerms(pod, node) ||
+				!predicates.NodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+				return
+			}
+			matchCount := 0
+			for _, existingPod := range nodeInfo.Pods() {
+				match, err := predicates.PodMatchesAllSpreadConstraints(existingPod, pod.Namespace, constraints)
+				if err != nil {
+					t.setError(err)
+					cancel()
+					return
+				}
+				if match {
+					matchCount++
+				}
+			}
+			// add matchCount up to EACH node which is at least in one topology domain
+			// with current node
+			for _, constraint := range constraints {
+				tpKey := constraint.TopologyKey
+				pair := topologyPair{key: tpKey, value: node.Labels[tpKey]}
+				for _, nodeName := range t.topologyPairToNodeNames[pair] {
+					atomic.AddInt64(t.counts[nodeName], int64(matchCount))
+					atomic.AddInt64(&t.total, int64(matchCount))
+				}
+			}
+		}
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+	if t.firstError != nil {
+		return nil, t.firstError
+	}
+
+	var maxCount, minCount int64
+	for _, node := range nodes {
+		if t.counts[node.Name] == nil {
+			continue
+		}
+		// reverse
+		count := t.total - *t.counts[node.Name]
+		if count > maxCount {
+			maxCount = count
+		} else if count < minCount {
+			minCount = count
+		}
+		t.counts[node.Name] = &count
+	}
+	// calculate final priority score for each node
+	// TODO(Huang-Wei): in alpha version, we keep the formula as simple as possible.
+	// current version ranks the nodes properly, but it doesn't take MaxSkew into
+	// consideration, we may come up with a better formula in the future.
+	maxMinDiff := maxCount - minCount
+	for i := range nodes {
+		node := nodes[i]
+		result[i].Host = node.Name
+
+		// debugging purpose: print the value for each node
+		// score must be pointer here, otherwise it's always 0
+		if klog.V(10) {
+			defer func(score *int, nodeName string) {
+				klog.Infof("%v -> %v: EvenPodsSpreadPriority, Score: (%d)", pod.Name, nodeName, *score)
+			}(&result[i].Score, node.Name)
+		}
+
+		if t.counts[node.Name] == nil {
+			result[i].Score = 0
+			continue
+		}
+		if maxMinDiff == 0 {
+			result[i].Score = schedulerapi.MaxPriority
+			continue
+		}
+		fScore := float64(schedulerapi.MaxPriority) * (float64(*t.counts[node.Name]-minCount) / float64(maxMinDiff))
+		// need to reverse b/c the more matching pods it has, the less qualified it is
+		// result[i].Score = schedulerapi.MaxPriority - int(fScore)
+		result[i].Score = int(fScore)
+	}
```

Comment:
this we can run in parallel as well.

Reply:
It seems like a not very expensive loop (we only iterate through nodes, not pods). It might require perf-testing.

Reply:
Probably, it depends on scale and cluster state though. Not worth doing it now though, I agree.